### PR TITLE
Various fixes to the live preview

### LINKED ISSUE #88: Can charites serve work properly on Linux?
Warm greeting! Thank you once again for the charites development!

After my work in my own PC, now I would like to share the styling work with my colleagues and use charites in Linux.
I have found some issue here. I would appreciate if if you could kindly confirm it!

**Describe the bug**
During charites serve, updated file is not detected automatically.

**To Reproduce**

* Install charites on Red Hat Enterprise Linux (release 8.4)
* run "charites serve"
* Open the Charites Live Preview through http://(ip address):8080
* update yaml files under layers
* Then, edit is not detected automatically. 
     * CLI does not show any message.
     * Charites Live Preview does not change.
     * (IF I type F5 on the browser, the edit is then reflected on Charites Live Preview, but no message in CLI..)

**Expected behavior**
Even in Linux, I expected that the update/edit of yml file is automatically detected and live-view automatically update.

**Screenshots**
2022-01-24-charites

**Charites Version:**
- Charites version: 0.X.X

**Desktop (please complete the following information):**

- Server OS: Red Hat Enterprise Linux (release 8.4)
- Browser: Google Chrome in Windows PC
(Charites is run on Linux server while I saw the preview from other PC.)

**Additional context**

### LINKED ISSUE #125: Refactor preview app.js files to reduce duplicated code
**Is your feature request related to a problem? Please describe.**
Right now, there are three main app.js files for each kind of preview environment, with a lot of duplicated code (because they all use mapbox-gl-js or maplibre-gl-js).

**Describe the solution you'd like**
The code would be a lot more maintainable if this code could be consolidated in to a single file.

**Describe alternatives you've considered**
Using the preview environment as-is, which is fine for now, but as more features are added to the preview environments, will become less maintainable.

**Additional context**
n/a

### LINKED ISSUE #128: LivePreview has some issue without center and zoom in the style.yml.
**Describe the bug**
Without center and zoom in the style.yml, we cannot get the live preview at "http://localhost:8080" because it is redirected to "http://localhost:8080/#NaN/NaN/NaN."

(We can access to the live preview by manually [http://localhost:8080/#0/0/0 .  )  

**To Reproduce**

1. Prepare a style yaml without center and zoom.
2. Run "charites serve"
3. Open  "http://localhost:8080" at your browser. (It is redirected to http://localhost:8080/#NaN/NaN/NaN)

**Expected behavior**
Because center and zoom are optional based on the style specification, if there is no information, we would like to use default value like 0/0/0 for z/x/y.

**Screenshots**
image

**Charites Version:**

- Charites version: 0.2.0

**Desktop (please complete the following information):**

- OS: [e.g. Windows] Docker on windows 10
- Browser [e.g. chrome, safari] Chrome

**Additional context**

### PR DESCRIPTION
**Pull request by keichan34**
## Description

Fixes #88, fixes #125, fixes #128, in addition to the issues below:

* The mapbox provider was broken because mapboxgl.Map was mistakenly replaced with maplibregl.Map
* Added automatic zoom/center to the Geolonia provider
* Add layer control to the Geolonia provider
* Throw an error if the Mapbox access token is not set when using the mapbox provider

## Type of Pull Request

- [ ] Adding a feature
- [x] Fixing a bug
- [ ] Maintaining documents
- [ ] Others ()

## Verify the followings

- [x] Code is up-to-date with the `main` branch
- [x] No build errors after `npm run build`
- [x] No lint errors after `npm run lint`
- [x] No errors on using `charites help` globally
- [x] Make sure all the exsiting features working well
- [ ] Have you added at least one unit test if you are going to add new feature?
- [ ] Have you updated documentation?

### CODE REVIEW

File: provider/default/shared.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+      // detect center & zoom from map style json
+      let center = mapStyleJson.hasOwnProperty('center')
+        ? mapStyleJson.center
+        : undefined
+      let zoom = mapStyleJson.hasOwnProperty('zoom')
+        ? mapStyleJson.zoom
+        : undefined
+
+      // detect center & zoom from tile json
+      if (center === undefined || zoom === undefined) {
```
Comment (smellman) [inline provider/default/shared.js:22]: I think it need failback if raster tile only pattern like this style.
https://maplibre.org/maplibre-gl-js-docs/example/3d-terrain/

**Comment by ubukawa** [comment]
I experienced "permission denied" when I run "convert" although I had a permission to make files in that directory. (both in docker and my linux environment.)
I did not experienced this before, but is it some how related to this pull request??

image


For example, in the Docker.
It is okay with the previous version.
image

But, if I uninstalled it and installed this version, it did not work...
image
image

**Comment by keichan34** [comment]
@ubukawa Sorry about this, I forgot to rebase on the current master after merging #121 , which fixes this issue. There's a separate issue where errors where the file has failed to write are being swallowed and replaced with a generic "permission denied" error, which I just reported in #130. It should be working now (I tested with your https://github.com/ubukawa/cmv-test/ repository on Mac -- I'll try again in Linux later today)

**Review by ubukawa — APPROVED** [review]
Thank you! 
I confirmed that charites worked well.

- The error I reported yesterday (permission denied) was solved. 
- legend plug-in appears in the live preview
- We can use 0/0/0 even the style.json does not have "center". 

For future consideration:
I realized that we may be able to use the latest version of maplibre gl js (2.4.0), if it does not affect others, as we use v1.15.2 now. https://github.com/unvt/charites/blob/main/provider/default/index.html#L5

The live preview worked well in the localhost, but when I run it in my webserver and accessed through the internet, it had some issue, maybe I guess it is  due to CORS issue. 
image

File: provider/default/shared.js (hunk 1 above)
Comment (smellman) [reply provider/default/shared.js:22]: resolved.

**Review by smellman — APPROVED** [review]

**Review by yuiseki — APPROVED** [review]
I've checked live preview of this branch has correctly worked on both of maplibre and mapbox.

I apologize for breaking the functionality with my previous pull request and I appreciate you correcting my failure.

LGTM!

**Review by naogify — CHANGES_REQUESTED** [review]
@keichan34 

Thank you! I confirmed the following functions works well!
- The mapbox provider was broken because mapboxgl.Map was mistakenly replaced with maplibregl.Map
- Add layer control to the Geolonia provider
- Throw an error if the Mapbox access token is not set when using the mapbox provider

> Added automatic zoom/center to the Geolonia provider

Mapbox GL JS / MapLibre GL JS have this function by default. So, I think we don't need to implement it by ourself. (If no center, zoom specified, it will move to 0/0/0)

https://github.com/unvt/charites/blob/e4d18ee60f3caa5b2636d84f72a51b567cb73743/provider/default/app.js#L8-L9

**Comment by keichan34** [comment]
@naogify 

> Added automatic zoom/center

This feature was added in #119, but only to default and mapbox providers -- I applied it to the geolonia provider as well. It's different from the default behavior because in addition of looking at the style, it also looks at the tile data sources.
The bug I fixed in relation to this feature was #128

**Review by naogify — APPROVED** [review]
@keichan34 
Oh, I see, #119 added the ability to zoom to the center of a tile, and there was a bug in that. I didn't follow it... Sorry.

**Pull request merged**